tools/substitute.sh

```diff
@@ -0,0 +1,41 @@
+#!/bin/sh
+# Shell script equivalent of Metamath SUBSTITUTE tool
+# Mario Carneiro 5-Jan-2022
+
+usage() {
+  cat << HELP
+SUBSTITUTE - Make a simple substitution on each line of the file
+
+This command performs a simple string substitution in each line of a file.
+If the string to be replaced is "\n", then every other line will
+be joined to the one below it.  If the replacement string is "\n", then
+each line will be split into two if there is a match.
+The MATCHSTR specifies a string that must also exist on a line
+before the substitution takes place; null means match any line.
+The OCCURRENCE is an integer (1 = first occurrence on each line, etc.)
+or A for all occurrences on each line.
+Syntax:  substitute.sh OLDSTR NEWSTR OCCURRENCE MATCHSTR < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" -o "$1" = "--help" ]; then usage; exit; fi
+if [ $# -ne 4 ]; then usage; exit 1; fi
+case "$3" in
+  A | a)  all=1;;
+  '' | '0' | *[!0-9]*)  usage; exit 1;;
+  *)  occs=$3;;
+esac
+
+# multiline oldStr not supported
+awk -v oldStr="$1" -v newStr="$2" -v matchStr="$4" \
+  -v all="$all" -v occs=$occs \
+'BEGIN {
+  gsub(/[][\\.^$(){}|*+?]/, "\\\\&", oldStr);
```

Comment:
the caret ^ needs to be escaped

Reply:
Only if it starts the bracket list.

Reply:
From gawk manual section 3.4 bracket expression p. 54
To include one of the characters ‘\’, ‘]’, ‘-’, or ‘^’ in a bracket expression, put a ‘\’ in
front of it.
The exceptional handling of ] at the beginning of the list is explicitly mentioned, but to my knowledge nothing extra is said about the ^.

So this might be technically correct code, but the contract is violated.  Which puts the code at risk, if other flavours of awk are used, or if future developments change the usage of the ^.